plugins/base/src/main/kotlin/transformers/pages/sourcelinks/SourceLinksTransformer.kt

```diff
@@ -2,83 +2,104 @@ package org.jetbrains.dokka.base.transformers.pages.sourcelinks
 
 import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiDocumentManager
-import com.intellij.psi.util.PsiTreeUtil
 import org.jetbrains.dokka.DokkaConfiguration
-import org.jetbrains.dokka.links.DRI
+import org.jetbrains.dokka.base.translators.documentables.PageContentBuilder
 import org.jetbrains.dokka.model.DescriptorDocumentableSource
+import org.jetbrains.dokka.model.DocumentableSource
+import org.jetbrains.dokka.model.PsiDocumentableSource
 import org.jetbrains.dokka.model.WithExpectActual
-import org.jetbrains.dokka.model.properties.PropertyContainer
 import org.jetbrains.dokka.pages.*
 import org.jetbrains.dokka.plugability.DokkaContext
 import org.jetbrains.dokka.transformers.pages.PageTransformer
 import org.jetbrains.kotlin.descriptors.DeclarationDescriptorWithSource
-import org.jetbrains.kotlin.kdoc.psi.api.KDoc
-import org.jetbrains.kotlin.psi.KtDeclaration
 import org.jetbrains.kotlin.resolve.source.getPsi
 import org.jetbrains.kotlin.utils.addToStdlib.cast
 import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
-class SourceLinksTransformer(val context: DokkaContext) : PageTransformer {
+class SourceLinksTransformer(val context: DokkaContext, val builder: PageContentBuilder) : PageTransformer {
+    private val sourceLinks = getSourceLinks()
 
-    override fun invoke(input: RootPageNode): RootPageNode {
+    override fun invoke(input: RootPageNode) =
+        input.transformContentPagesTree { node ->
+            when(val documentable = node.documentable){
+                is WithExpectActual -> resolveSources(documentable)
+                    .takeIf{ it.isNotEmpty() }
+                    ?.let { node.addSourcesContent(it) }
+                    ?: node
+                else -> node
+            }
+        }
 
-        val sourceLinks = context.configuration.passesConfigurations
-            .flatMap { it.sourceLinks.map { sl -> SourceLink(sl, it.platformData) } }
+    private fun getSourceLinks() = context.configuration.passesConfigurations
+        .flatMap { it.sourceLinks.map { sl -> SourceLink(sl, it.platformData) } }
 
-        return input.transformContentPagesTree { node ->
-            node.documentable.safeAs<WithExpectActual>()?.sources?.map?.entries?.fold(node) { acc, entry ->
-                sourceLinks.find { entry.value.path.contains(it.path) && it.platformData == entry.key }?.run {
-                    acc.modified(
-                        content = acc.content.addSource(
-                            entry.key,
-                            entry.value.cast<DescriptorDocumentableSource>().toLink(this)
-                        )
-                    )
-                } ?: acc
-            } ?: node
+    private fun resolveSources(documentable: WithExpectActual) = documentable.sources.map.entries
```

Comment:
I think you can remove the `.map.entries` part entirely

Reply:
Done